Fix string handling

kariya-mitsuru opened the pull request:
1. string cannot contain control characters.
2. hex escape like "\80" represents a byte, not a unicode character.

Passed new 250 tests.

**before**
```
End ".../wain/spec-test/wasm-testsuite/names.wast":
  total: 487, passed: 411, failed: 76, skipped: 0

End ".../wain/spec-test/wasm-testsuite/utf8-invalid-encoding.wast":
  total: 177, passed: 1, failed: 176, skipped: 0

Results of 76 files:
  total: 19740, passed: 18796, failed: 748, skipped: 196
```

**after**
```
End ".../wain/spec-test/wasm-testsuite/names.wast":
  total: 487, passed: 485, failed: 2, skipped: 0

End ".../wain/spec-test/wasm-testsuite/utf8-invalid-encoding.wast":
  total: 177, passed: 177, failed: 0, skipped: 0

Results of 76 files:
  total: 19740, passed: 19046, failed: 498, skipped: 196
```

kariya-mitsuru committed:
Fix string handling

1. string cannot contain control characters.
2. hex escape like "\80" represents a byte, not a unicode character.

rhysd commented:
Thank you! Let me check this tomorrow.

rhysd commented on wain-syntax-text/src/lexer.rs:
```diff
@@ -236,14 +243,70 @@ impl<'s> Lexer<'s> {
-            } else if c == '"' {
-                let token = Token::String(&self.source[start + 1..offset]); // + 1 consumes '"'
-                return Ok(Some((token, start)));
+        let mut buf = vec![];
```
リテラルの中のパースを parser から lexer に移したのって，spec test runner と wain-syntax-text で実装を共有したかったからですよね？lexer はソースをトークンに区切るのが仕事で，リテラルの中身をパースするのは parser 側でやるべきという判断で元々 parser 側に実装していて，Lexer は余計なアロケーションをせず，極力トークンを区切ることに徹したいです．

下記のような実装でどうでしょう？

1. `parse_escaped_text` のバグを修正
2. `parse_escaped_text` を `parse_string_literal_content` にリネーム（元々 internal 用であまり適切でない名前なので…）
3. `parse_string_literal_content` を外に公開
4. `wain_syntax_text::parser::Parser` と spec test の parser の両方から import して使う

もし大変そうであれば，一旦この PR はマージして，僕の方で書き換えてしまっても OK なので，遠慮せず言ってください

kariya-mitsuru replied:
> リテラルの中のパースを parser から lexer に移したのって，spec test runner と wain-syntax-text で実装を共有したかったからですよね？

それもありますが、そもそも wasm の spec でエスケープシーケンス等の解釈は 6 章の Text Format の中なので、lexer で処理するのが適切と考えたからです。
実際、spec のリポジトリにある OCaml の参照実装でも lexer で処理しています。

また、lexer.rs で
```
    String(&'source str), // Should parse the literal into Vec<u8>?
```
と
```
        // Note: Content of literal will be parsed in parser so we need to traverse the content
        // twice. It would be better to parse the content into Vec<u8> here.
```
とあったのも、lexer に移した方が良いと思った理由の一つです。
（確かに速度的にも有利かなぁ、と）

rhysd replied:
> wasm の spec でエスケープシーケンス等の解釈は 6 章の Text Format の中なので、lexer で処理するのが適切と考えたからです。

Text Format の章は Wasm をテキストで書く際の構文を定義しているので、lex と parse 両方の処理で必要なことを表しています。なので Text Format の章に書かれている = lex でやるべき とはならないのではと思います。

> とあったのも、lexer に移した方が良いと思った理由の一つです。

（完全に書いた記憶が残ってない…）自分で書いておいてアレですが、それを言うと float literal も2回舐めてるわけですしね…

とりあえず現状はこれで OK にして、そのあとどうするかはちょっと考えてみます。もう1つのほうのコメントご確認お願いします。

kariya-mitsuru replied:
> Text Format の章は Wasm をテキストで書く際の構文を定義しているので、lex と parse 両方の処理で必要なことを表しています。なので Text Format の章に書かれている = lex でやるべき とはならないのではと思います。

確かにその通りですね。
少なくとも 6 章全体を lexer で処理すべきではないし、そもそもできないですね。
何で自分がこの処理を lexer に書いた方がいいと思ったかと言うと、

- `string` と言う非終端記号はトークンである。
- `string` と言う非終端記号には正しいエスケープシーケンスしか含まれていない。（逆に言えば、正しくないエスケープシーケンスが含まれたものは `string` と言う非終端記号ではない）
- どうせ正しいエスケープシーケンスかどうかをチェックするのであればその場で `Vec<u8>` に変換してしまった方が二度手間を省けて楽。

だからですね。
これが 6 章に書かれてるから 6 章全体を lexer でと勘違いしましたが、全然違いましたね…

> （完全に書いた記憶が残ってない…）自分で書いておいてアレですが、それを言うと float literal も2回舐めてるわけですしね…

float literal は（も？）悩ましいですね。
そもそも `f32` と `f64` の区別が lexer の段階では付かないので…
少なくとも指数部ぐらいは `u32` に変換しちゃってもいいかなぁ、と眺めながら思ってました…

で、今文法見てて思ったんですが、これ `fnn.const` に整数渡せないですね。
エラーとするのが正しいのかも…
ちなみに wasm 1.1 では `float` と `hexfloat` の定義に小数点も指数部も必須じゃなくなったので、そもそも整数リテラルと浮動小数点リテラルの区別すら付かなくなったので渡せるようになっただけですね…
（いわゆる defect かな…）

rhysd replied:
なるほど．正しいエスケープシーケンスが含まれてない文字列リテラルは文字列リテラルではなくなって，reserved あたりになるという話ですね．確かにどうせ舐めるならパースもしておきたいというのはそのとおりです（文字列リテラルは巨大になりがちですし）．

> 少なくとも指数部ぐらいは u32 に変換しちゃってもいいかなぁ、と眺めながら思ってました…

うーん，パースの処理はできるだけパーサ側に寄せたいですね．部分的にこっちでやって残りはあっちでやってというのはなるべく避けたいです．浮動小数点数の各パーツに分解する現状で良いのかなと思ってます．巨大になりうる文字列リテラルに比べて浮動小数点数リテラルはせいぜい数十バイトなので，オーバーヘッドも極小でしょうし．

> で、今文法見てて思ったんですが、これ fnn.const に整数渡せないですね。

うーん，仕様を見る限りでは渡せるようにみえます．まず

https://webassembly.github.io/spec/core/text/instructions.html

を見ると `fN.const` の引数は `fN` です．それで，`fN` の定義は符号と `fNmag` で，`fNmag` は `float` と `hexfloat` と NaN と inf です．

https://webassembly.github.io/spec/core/text/values.html#text-float

それで，`float` は

<img width="651" alt="スクリーンショット 2020-05-25 20 43 49" src="https://user-images.githubusercontent.com/823277/82810061-82016a00-9ec8-11ea-8744-b06774f46370.png">

の1つの目の branch で `.` を省略できます．また `hexfloat` は

<img width="643" alt="スクリーンショット 2020-05-25 20 44 57" src="https://user-images.githubusercontent.com/823277/82810112-9e050b80-9ec8-11ea-8977-70923a4d09b0.png">

の1つ目の branch で `.` を省略できます．これらの場合，`fN.const` の引数は整数値になっています．

kariya-mitsuru replied:
このサイトは既に 1.0 の仕様からずれてますよね…（1.1 のドラフトになってる）
W3C の Recommendation のサイトでは `.` の省略が無いです。
https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#floating-point%E2%91%A6

で、良く分からないのが、Wasm のリポジトリのタグ `wg_1.0` ではやはり `.` の省略が無いんですが、ブランチ `w3c-1.0` ではなぜか `.` の省略があるんですよね…

ちなみに、この修正は過去に Issue が上がってました。
https://github.com/WebAssembly/spec/issues/1066

仕様のバグ扱いで受け入れるのが正解ですかね…

rhysd replied:
あっ，なるほど．そういう経緯でしたか… すみません，そこまでは把握してませんでした．貼っていただいたリンクの issue の修正を見る限り，整数も受け入れる今の実装で良さそうですね

rhysd commented on spec-test/src/parser.rs:
```diff
@@ -287,8 +210,8 @@ impl<'s> Parse<'s> for EmbeddedModule {
-                        Some(Token::String(s)) => {
-                            bin.append(&mut parser.parse_escaped(s)?);
+                        Some(Token::String(ref s, _)) => {
+                            bin.extend_from_slice(s);
```
`consume()` はトークンを値で返すので，`ref` は不要のはずです．

```rust
Some(Token::String(bytes, _)) => {
    bin.extend_from_slice(bytes);
}
```

`bytes` を `Vec<u8>` でムーブすれば，

```rust
fn parse_escaped_text(&mut self, bytes: Vec<u8>) -> Result<'s, String> {
    String::from_utf8(bytes).map_err(|e| self.fail(ParseKind::Utf8Error(e)))
}
```

とできて，文字列リテラルのバイト列のコピーを1回（`.to_string` 分）減らせると思います．

rhysd commented on spec-test/src/parser.rs:
```diff
@@ -270,7 +193,7 @@ impl<'s> Parse<'s> for EmbeddedModule {
                     match parser.consume()? {
-                        Some(Token::String(s)) => {
+                        Some(Token::String(ref s, _)) => {
                             text.push_str(&parser.parse_escaped_text(s)?);
```
`consume()` はトークンを値で返すので，`ref` は不要のはずです．

```rust
Some(Token::String(bytes, _)) => {
    text.push_str(&parser.parse_escaped_text(bytes)?);
}
```

`bytes` を `Vec<u8>` でムーブすれば，

```rust
fn parse_escaped_text(&mut self, bytes: Vec<u8>) -> Result<'s, String> {
    String::from_utf8(bytes).map_err(|e| self.fail(ParseKind::Utf8Error(e)))
}
```

とできて，文字列リテラルのバイト列のコピーを1回（`.to_string` 分）減らせると思います．

kariya-mitsuru replied:
なるほど、`ref` は削除します。
ちなみに `str::from_utf8` はコピーが発生しませんが `String::from_utf8` はコピーが発生するので、コピーの回数は変わらないです。

rhysd replied:
いえ，`String::from_utf8` はコピー発生しないです．

https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8 より

> This method will take care to not copy the vector, for efficiency's sake.

やってるのは，引数の `Vec` をムーブで受け取って UTF-8 validation やって `String` の中に持つだけです（実装的には `String` は `Vec<u8>` を wrap しただけの構造体になってます）

kariya-mitsuru replied:
おっ、ホンマや…

rhysd reviewed (changes requested):
（コメントをつけるところを間違えていたので、再コメントしました）

rhysd commented on wain-syntax-text/src/lexer.rs:
```diff
@@ -633,15 +710,27 @@ mod tests {
 
     #[test]
     fn strings() {
```
`assert_lex_string` マクロはこの関数の中でしか使われてないので、ここで定義お願いします

rhysd commented on wain-syntax-text/src/lexer.rs:
```diff
@@ -633,15 +710,27 @@ mod tests {
+        assert_lex_error!(r#""\u{hello!}""#, LexErrorKind::InvalidStringFormat);
+        assert_lex_error!(r#""\u{d800}""#, LexErrorKind::InvalidStringFormat);
+        assert_lex_error!(r#""\u{dfff}""#, LexErrorKind::InvalidStringFormat);
+        assert_lex_error!(r#""\u{110000}""#, LexErrorKind::InvalidStringFormat);
```
https://github.com/rhysd/wain/pull/13/files#diff-26164d7624c8205d9952301747d56114R303 を通る場合のテストもお願いします

kariya-mitsuru committed:
Follow Github comments

remove unnecessary `ref` when match Token::String
move the assert_lex_string macro into the function string

rhysd commented:
コードは LGTM です．あとは https://github.com/rhysd/wain/pull/13#discussion_r429891508 コメントの確認だけお願いします

rhysd commented:
ちょっと 1.1 対応をどうするかは考えないとですね… コードコメント内の URL は最新のドラフトを見ちゃってますし
